Porting the fixes/hacks from the original IQSS fork

landreev opened the pull request:
These are the changes from the local fork that were needed to get this thing to work with the Dataverse model of exporting and storing metadata. 
I'd like to achieve an intermediate result of being able to build Dataverse from the current develop branch with these new gdcc/xoai libraries and have all the OAI parts work exactly as they are working now, when built with the old jars in local_lib. Before we proceed with any further cleanup and refactoring. 

I'll try to explain what these changes do and why that model was chosen. 

ref. #2

landreev committed:
The few extra fixes/hacks from the original IQSS fork of the lyncode v4 distribution that were needed to get this thing to work with the dataverse model exporting and storing metadata. (#2)

landreev commented:
(I'll fix the failing test tomorrow)

landreev committed:
_tentatively_ putting back the format check line (added by DSpace in 4.20). This fixes the failing test in GetRecordHandlerTest; but I will need to confirm that it does not interfere with the Dataverse use case. #2 (will explain this in more detail in the github issue)

poikilotherm commented on xoai-data-provider/src/main/java/io/gdcc/xoai/dataprovider/handlers/ListIdentifiersHandler.java:
```diff
@@ -69,8 +70,10 @@ else if (parameters.hasFrom() && parameters.hasUntil())
         } else {
-            if (!getRepository().getSetRepository().exists(parameters.getSet()) && !getContext().hasSet(parameters.getSet()))
-                throw new NoMatchesException();
+            /* TBH, I don't remember why the 2 lines below are commented out in 
+               the original IQSS fork. Will check. -- L.A. */
+            /*if (!getRepository().getSetRepository().exists(parameters.getSet()) && !getContext().hasSet(parameters.getSet()))
+                throw new NoMatchesException();*/
```
I *think* there might have been a validation check that failed, looking at https://github.com/IQSS/dataverse/issues/3307.

The very same problem seems to be one of the causes for the existance of [XlistRecordsHandler.java](https://github.com/IQSS/dataverse/blob/50f093be5bf14eb940a0807e9c36c7e6b1806e4a/src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/XlistRecordsHandler.java#L86) (the rest of the code in there seems to be mostly the same as the original [ListRecordsHandler.java](https://github.com/gdcc/xoai/blob/branch-5.0/xoai-data-provider/src/main/java/io/gdcc/xoai/dataprovider/handlers/ListRecordsHandler.java)...)

I actually found a bug that might have caused the trouble leading to the workaround: 
https://github.com/DSpace/xoai/blob/c8857979d809938250b02f98087b2f14e5c24f51/xoai-data-provider/src/main/java/io/gdcc/xoai/dataprovider/handlers/helpers/SetRepositoryHelper.java#L64-L67

(Uses the wrong object for the comparison - should have been `setSpec`, **not** `set`).

I fixed this with https://github.com/gdcc/xoai/commit/149ac89f928df470793f62016d55ef4475cc28df and https://github.com/gdcc/xoai/commit/a95172d8472df75b96ad1b365116737e8415a822, implementing `Set.equals()` and using `List.contains()`, so I *think* we should be able to fix this one, too.

poikilotherm commented on xoai-data-provider/src/main/java/io/gdcc/xoai/dataprovider/handlers/GetRecordHandler.java:
```diff
@@ -98,6 +97,7 @@ public GetRecord handle(OAICompiledRequest parameters) throws OAIException, Hand
                 for (About about : item.getAbout())
                     record.withAbout(about);
             }
+            */
```
Do you remember *why* you attached the "raw" metadata instead of the transformed version?

poikilotherm commented on xoai-common/src/main/java/io/gdcc/xoai/model/oaipmh/GetRecord.java:
```diff
@@ -31,6 +31,13 @@ public void write(XmlWriter writer) throws XmlWriteException {
         }
     }
 
+    /* Not to be confused with the similarly-named constructor, this
+       method exposes the "private final Record record" to other parts
+       of the OAI framework: */ 
+    public Record getRecord() {
+        return record;
+    }
+    
```
The calling graph for this in Dataverse is: `OAIServlet.processRequest()` -> `OAIServlet.writeGetRecord()` -> `XgetRecord.writeToStream()` -> `GetRecord.getRecord()`.

~~I have no idea why the orignal library intended way wasn't used.~~ (That would be an `XmlWriter` with the `OutputStream` of the servlet, creating the `GetRecord` verb and sending it down the path of an `OAIPMH.write()`.) This is code back from 2016, so it might remain unknown. EDIT: I see the purpose of `Xrecord` being capable of adding Dataverse metadata. Let's see how we can do this in a cleaner way (decorator pattern?).

For the way forward: I propose we cleanup the Dataverse side once we have a XOAI-snapshot available, see what we really need and incorporate those changes in this library. Lots of classes had been forked with minimal changes and it's hard to say why all these things happened as they did...

landreev replied:
I'll answer/explain this in #2 (or here) tomorrow.

landreev commented:
@poikilotherm 
This is all based on my best recollection of what we did back in 2016 and why those decisions were made:
We incorporated the basic XOAI implementation framework of managing the sets, records and service verbs. However, we changed quite radically the way the actual metadata content is served. XOAI treats the metadata fragments in the repository as "live" XML. It is parsed; it can be transformed into another format, if needed; and the entire OAI output is written as an XML stream. It was clearly implemented under the assumption that these metadata fragments are fairly small, like a typical DC record; so their Metadata objects have these records fully loaded in memory. The 2 OAI verbs that serve the metadata content, GetRecord and ListRecords, is structured as an outer XML record formatted under OAI's own schema, serving as a wrapper around the dataset metadata proper, with its own schema specified, and its own metadata prefix. 
[I'm spelling this out in painful detail under the assumption that somebody who doesn't know much about OAI yet may need to read this as well]
For example, this is the outer wrapper formatting of the GetRecord output: 
```
<?xml version="1.0" encoding="UTF-8"?> 
<OAI-PMH xmlns="http://www.openarchives.org/OAI/2.0/" 
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/
         http://www.openarchives.org/OAI/2.0/OAI-PMH.xsd">
  <responseDate>xxx</responseDate>
  <request verb="GetRecord" identifier="..."
           metadataPrefix="oai_dc">https://.../oai</request>
  <GetRecord>
   <record> 
    <header>
      <identifier>...</identifier> 
      <datestamp>...</datestamp>
      <setSpec>...</setSpec> 
      <setSpec>...</setSpec>
    </header>
    <metadata>
      [ ... actual dataset metadata inserted here ... ]
    </metadata>
  </record>
 </GetRecord>
</OAI-PMH>
```
So a complete output example would look like this: 

```
<OAI-PMH xmlns="http://www.openarchives.org/OAI/2.0/" 
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
        xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/ 
        http://www.openarchives.org/OAI/2.0/OAI-PMH.xsd">
  <responseDate>2022-05-10T17:42:06Z</responseDate>
  <request verb="GetRecord" identifier="doi:10.7910/DVN/QTHFVS" 
           metadataPrefix="oai_dc">https://dataverse.harvard.edu/oai</request>
  <GetRecord>
    <record>
    <header>
      <identifier>doi:10.7910/DVN/QTHFVS</identifier>
      <datestamp>2022-05-10T06:08:02Z</datestamp>
      <setSpec>IQSS</setSpec>
    </header>
    <metadata>
      <oai_dc:dc xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
            xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/" 
            xmlns:dc="http://purl.org. /dc/elements/1.1/" 
            xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc/ 
            http://www.openarchives.org/OAI/2.0/oai_dc.xsd">
        <dc:title>Input parameter file for a global dynamo simulation of Saturn with an SSL</dc:title>
        <dc:identifier>https://doi.org/10.7910/DVN/QTHFVS</dc:identifier>
        <dc:creator>Yadav, Rakesh</dc:creator>
        <dc:publisher>Harvard Dataverse</dc:publisher>
        <dc:description>The data set contains the input parameter file ...</dc:description>
        <dc:subject>Astronomy and Astrophysics</dc:subject>
      </oai_dc:dc>
    </metadata>
  </record>
  </GetRecord>
</OAI-PMH>
```

(ListRecords is basically the same thing, but for multiple records/datasets)

In the XOAI implementation, in order to produce the record above, we use an XmlWriter to write the service header; then get the oai_dc format record from the repository (let's assume it's already available, no transformations needed), load it in memory in full, parse it and output it; then finish writing the closing elements of the outer service XML. 

We made a decision to avoid doing anything potentially resource-expensive inside the Dataverse OAI framework. No metadata transformations would be done in there; meaning, only the datasets for which the metadata in the OAI formats have already been exported and cached would be added to the OAI sets and served as OAI records. We also decided to avoid having to load an entire metadata record into memory at once. Instead the XOAI class would hold onto an InputStream pointing to the cached metadata export, which it would stream directly to the output once the service header part of the OAI response has been written. At that time we were operating under the assumption that Dataverses would continue harvesting **FULL** DDI records from each other ("full DDI", meaning including the data/variable-level sections; so, for some of our datasets with lots of tabular files, the full DDI exports were running into tens of MBs; too large to keep in memory in full; and definitely impractical to parse on the fly). Most of the modifications that were needed in order to implement these changes were done on the Dataverse side, in the classes overwriting the corresponding XOAI classes. In a couple of cases the hacks above were necessary (such as in the case of needing to expose the private variable, above). At that point we were in a hurry and behind schedule to add harvesting to Dataverse, hence all the questionable choices that were made regarding the quality of the hacks. 

It's worth pointing out that we have switched to a model where we only serve the "light DDI" format under OAI, i.e. only the dataset-level metadata without the data sections (the format is called "oai_ddi" in Dataverse; the full format, "ddi" is available for download, but not to OAI clients). These records are much lighter; but can still be quite large. And I still like the idea of keeping the OAI framework super lightweight and relying on the metadata records having already been exported and cached externally. I do of course support the idea of achieving this in a cleaner way. 

I hope this makes sense; and I hope my own understanding of it all is correct, after the eons that have passed since those ancient times.

poikilotherm commented:
Thank you @landreev for this great explanation on the backgrounds. With this intel, I could dig through the existing Dataverse OAI data provider come with a lot more ease and understand what is going on!

My findings:
- You are retrieving the cached metadata from the file storage, saved with the datasets files as auxiliary files ([start here](http://github.com/IQSS/dataverse/blob/develop/src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/Xrecord.java#L89-L89))
- You are reusing lots of the XOAI classes for headers, the OAIPMH root handle etc., but wrap it in a custom serialization to String and then write these to output stream
- You removed the metadata formatting in the `GetRecordHandler` class within the forked library and removed the same from the `ListRecordHandler` by cloning the implementation within the Dataverse codebase and adapting as needed.

I'm taking the liberty of proposing a different approach. How about:
- Extend the [`Metadata`](http://github.com/gdcc/xoai/blob/bc9766aef87d54993b9f08d4e0aa890eb7f841a5/xoai-common/src/main/java/io/gdcc/xoai/model/oaipmh/Metadata.java#L43-L43) class to associate `Metadata(InputStream)` with the stream only - opening, reading and closing it when it's `Metadata.write(XmlWriter)` time (so not read all into memory)
- Refactor the `GetRecordHandler` and `ListRecordHandler` not to pass an `InputStream` from the `XSLPipeline` but a `String` (which is cleaner design anyway)
- Refactoring our `ItemRepository` implementation to inject the auxiliary file input stream with the `Metadata` injected into the `Item`
- (Optionally extend the `Item` interface with a `hasMetadata()` method for eased presence check)
- Extending the `GetRecordHandler` and `ListRecordsHandler` to use Item.getMetadata(), reusing an existing one instead of creating a new `Metadata` instance when present (which sends things down the pipe, which we don't want) - just like it had been done in the forked library version, but keeping both ways.

This should a) avoid building the metadata from scratch everytime and b) avoid reading the metadata files until the very last moment when the writer gets activated.

I propose to discuss this via videoconference - much easier to follow the code pathes and see the things connect.

landreev commented:
Closing this one.
